blackduck/ClientCore.py

```diff
@@ -1,111 +0,0 @@
-
-def _request(
-    self, 
-    method,
-    url,
-    name='',
-    parameters=[],
-    **kwargs
-    ):
-    """[summary]
-
-    Args:
-        method ([type]): [description]
-        url ([type]): [description]
-        name (str, optional): name of the reqested resource. Defaults to ''.
-
-    Raises:
-        connect_timeout: often indicative of proxy misconfig
-        read_timeout: often indicative of slow connection
-
-    Returns:
-        json/dict/list: requested object, json decoded.
-    """
-
-    headers = {
-        'accept': 'application/json',
-        'content-type': 'application/json'
-    }
-    headers.update(kwargs.pop('headers', dict()))
-
-    if parameters:
-        url += self.get_parameter_string(parameters)
-
-    try:
-        response = self.session.request(
-            method=method, 
-            url=url, 
-            headers=headers,
-            **kwargs
-        )
-
-        if response.status_code / 100 != 2:
-            self.http_exception_handler(
-                response=response,
-                name=name
-            )
-
-        if 'Content-Type' in response.headers and 'internal' in response.headers['Content-Type']:
-            logging.warning("Response contains internal proprietary Content-Type: " + response.headers['Content-Type'])
-
-        response_json = response.json()
-
-    # Do not handle exceptions - just just more details as to possible causes
-    # Thus we do not catch a JsonDecodeError here even though it may occur
-    except requests.exceptions.ConnectTimeout:
-        logger.critical("could not establish a connection; this may be indicative of proxy misconfiguration")
-        raise
-    except requests.exceptions.ReadTimeout:
-        logger.critical("slow or unstable connection, consider increasing timeout")
-        raise
-    else:
-        return response_json
-    
-def _get_items(self, url, method='GET', page_size=100, name='', **kwargs):
-    """Utility method to get 'pages' of items
-
-    Args:
-        url (str): [description]
-        method (str, optional): [description]. Defaults to 'GET'.
-        page_size (int, optional): [description]. Defaults to 100.
-        name (str, optional): [description]. Defaults to ''.
-
-    Yields:
-        generator(dict/json): of items
-    """
-    offset = 0
-    params = kwargs.pop('params', dict())
-
-    while True:
-        params.update({'offset':f"{offset}", 'limit':f"{page_size}"})
-        items = self._request(
-            method=method,
-            url=url,
-            params=params,
-            name=name,
-            **kwargs
-        ).get('items', list())
-
-        for item in items:
-            yield item
-
-        if len(items) < page_size:
-            # This will be true if there are no more 'pages' to view
-            break
-
-        offset += page_size     
-
-def get_parameter_string(self, parameters=list()):
-    return '?' + '&'.join(parameters) if parameters else ''
```

Comment:
The requests library has named argument 'params' that should be used instead of 'parameters'

Reply:
Some blackduck endpoints have filters that reuse the same param name over and over, which isn't supported by the regular params dict (since dict keys are unique).

Reply:
Ugh, okay. Can you point me to one that I can play around with? I'll put this back in the meantime.

Reply:
I'll try and find it but iirc it was one of the vulnerability endpoints  `filter=securityRisk%3Amedium&filter=securityRisk%3Ahigh` although presumably any endpoint supporting multiple filters would have this problem.

Reply:
Okay I'll take a look and try out some things to see what works best. It was looking for a moment that we could get rid of _request / checked_request altogether...

Reply:
Turns out you can get the desired behavior in the following manner:
    params = {
        'name': ['foo', 'bar', 'baz'],
    }
    bd.session.get("/api/projects", params=params)
    # constructs url: /api/projects?name=foo&name=bar&name=baz

I'll document this so that end-users are aware.